### PR TITLE
executor: show operators' concurrency information in results of `EXPLAIN ANALYZE`

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve? 
Add the operator's concurrency information in `execution info` when we use the `EXPLAIN ANALYZE` statement.

After this PR, you can see:
```
mysql> explain analyze select avg(t1.a) from t1, t2 where t1.b = t2.b group by t1.a;
+------------------------------+-------+-----------+------------------------------------------------------------------+----------------------------------------------------------------------------+-------------+
| id                           | count | task      | operator info                                                    | execution info                                                             | memory      |
+------------------------------+-------+-----------+------------------------------------------------------------------+----------------------------------------------------------------------------+-------------+
| HashAgg_11                   | 7.19  | root      | group by:Column#9, funcs:avg(Column#8)                           | time:825.419µs, loops:2, rows:3, PartialConcurrency:4, FinalConcurrency:4  | N/A         |
| └─Projection_20              | 11.24 | root      | cast(Column#1), Column#1                                         | time:605.521µs, loops:2, rows:27, Concurrency:OFF                          | N/A         |
|   └─HashLeftJoin_12          | 11.24 | root      | inner join, inner:TableReader_19, equal:[eq(Column#2, Column#5)] | time:568.983µs, loops:2, rows:27, Concurrency:5                            | 8.234375 KB |
|     ├─TableReader_16         | 8.99  | root      | data:Selection_15                                                | time:373.189µs, loops:2, rows:9, rpc num:1, time:452.705µs                 | 284 Bytes   |
|     │ └─Selection_15         | 8.99  | cop[tikv] | not(isnull(Column#2))                                            | time:236.146µs, loops:10, rows:9                                           | N/A         |
|     │   └─TableScan_14       | 9.00  | cop[tikv] | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo      | time:216.908µs, loops:10, rows:9                                           | N/A         |
|     └─TableReader_19         | 8.99  | root      | data:Selection_18                                                | time:247.64µs, loops:2, rows:9, rpc num:1, time:307.705µs                  | 201 Bytes   |
|       └─Selection_18         | 8.99  | cop[tikv] | not(isnull(Column#5))                                            | time:173.387µs, loops:10, rows:9                                           | N/A         |
|         └─TableScan_17       | 9.00  | cop[tikv] | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo      | time:163.094µs, loops:10, rows:9                                           | N/A         |
+------------------------------+-------+-----------+------------------------------------------------------------------+----------------------------------------------------------------------------+-------------+
```

### What is changed and how it works?
Some operators including projection/ hashAgg/ hashJoin/ Index_lookup_Join/ index_merge_join/ index_hash_join , can be executed concurrently.  We add their concurrency information to the result when we execute the "explain analyze" statement. Besides, if the root task can't execute concurrently, the concurrency number will show ```Concurrency=OFF```.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: planner/core/explain.go [hunk 1]
```diff
@@ -172,6 +172,10 @@ func (p *PhysicalSelection) ExplainInfo() string {
 
 // ExplainInfo implements Plan interface.
 func (p *PhysicalProjection) ExplainInfo() string {
+	if p.Concurrency > 0 {
```
Comment (lamxTyler) [inline planner/core/explain.go]: I remember that the concurrency of projection worker is also controlled by row count, see https://github.com/pingcap/tidb/blob/master/executor/builder.go#L1194. Do we need to consider this?

Comment (Reminiscent) [reply planner/core/explain.go]: Yes, we need to record the concurrency information of the operator during the execution phase. Update. PTAL. @lamxTyler

File: util/execdetails/execdetails.go [hunk 2]
```diff
@@ -410,6 +412,16 @@ func (e *RuntimeStats) SetRowNum(rowNum int64) {
 	atomic.StoreInt64(&e.rows, rowNum)
 }
 
+// SetConcurrencyInfo sets the concurrency information.
+func (e *RuntimeStats) SetConcurrencyInfo(concurrencyInfo string) {
+	var val atomic.Value
+	val.Store(concurrencyInfo)
+	e.concurrency = val.Load().(string)
```
Comment (lamxTyler) [inline util/execdetails/execdetails.go]: It cannot prevent data race...

File: executor/aggregate.go [hunk 3]
```diff
@@ -224,6 +225,13 @@ func (e *HashAggExec) Close() error {
 	for range e.finalOutputCh {
 	}
 	e.executed = false
+
+	// len(e.partialWorkers) == len(e.finalWorkers)
+	concurrency := len(e.partialWorkers)
```
Comment (lamxTyler) [inline executor/aggregate.go]: Why don't we set it in planner phase like previous version? The concurrency can be caculated in planner.

Comment (Reminiscent) [reply executor/aggregate.go]: Because of the ```projection  concurrency``` can not be calculated accurately in the planner phase. See [here](https://github.com/pingcap/tidb/blob/fea5ce1114c43d1ea4795c7eedfa209c4610b76a/executor/projection.go#L33-L39) and [here](https://github.com/pingcap/tidb/blob/fea5ce1114c43d1ea4795c7eedfa209c4610b76a/executor/projection.go#L87-L92) for more details.

File: executor/aggregate.go [hunk 4]
```diff
@@ -224,6 +225,13 @@ func (e *HashAggExec) Close() error {
 	for range e.finalOutputCh {
 	}
 	e.executed = false
+
+	// len(e.partialWorkers) == len(e.finalWorkers)
```
Comment (SunRunAway) [inline executor/aggregate.go]: How cames this assertion?

See https://github.com/pingcap/tidb/blob/5c521c8e43a6706e3f98ed62b7d6544403bb932d/executor/aggregate.go#L254-L255

File: executor/aggregate.go [hunk 5]
```diff
@@ -224,6 +225,13 @@ func (e *HashAggExec) Close() error {
...
+	concurrency := len(e.partialWorkers)
+	if e.runtimeStats != nil && concurrency > 0 {
```
Comment (SunRunAway) [inline executor/aggregate.go]: Is `concurrency` possible to be negative?

File: executor/projection.go [hunk 6]
```diff
@@ -259,6 +259,10 @@ func (e *ProjectionExec) Close() error {
 		}
 		e.outputCh = nil
 	}
+	if e.runtimeStats != nil && e.numWorkers > 0 {
```
Comment (SunRunAway) [inline executor/projection.go]: I prefer making it specific when the executor can not run concurrently.

File: util/execdetails/execdetails.go [hunk 7]
```diff
@@ -410,6 +412,16 @@ func (e *RuntimeStats) SetRowNum(rowNum int64) {
...
+}
+
 func (e *RuntimeStats) String() string {
+	if len(e.concurrency) > 0 {
```
Comment (SunRunAway) [inline util/execdetails/execdetails.go]: I prefer making it specific when the executor can not run concurrently.

**Comment by SunRunAway** [comment]
Could you also do it in Index merge join & Index lookup join

File: util/execdetails/execdetails.go [hunk 8]
```diff
@@ -326,6 +326,8 @@ type RuntimeStats struct {
 	consume int64
 	// executor return row count.
 	rows int64
+	// executor concurrency information
+	concurrency string
```
Comment (lzmhhh123) [inline util/execdetails/execdetails.go]: Maybe it's better to set its type as int instead of the string.

File: util/execdetails/execdetails.go [hunk 9]
```diff
@@ -245,7 +245,7 @@ func (crs *CopRuntimeStats) RecordOneCopTask(address string, summary *tipb.Execu
 	crs.Lock()
 	defer crs.Unlock()
 	crs.stats[address] = append(crs.stats[address],
-		&RuntimeStats{int32(*summary.NumIterations), int64(*summary.TimeProcessedNs), int64(*summary.NumProducedRows)})
+		&RuntimeStats{int32(*summary.NumIterations), int64(*summary.TimeProcessedNs), int64(*summary.NumProducedRows), ""})
```
Comment (lzmhhh123) [inline util/execdetails/execdetails.go]: Unnecessary change.

File: util/execdetails/execdetails.go (hunk 8 above)
Comment (Reminiscent) [reply util/execdetails/execdetails.go]: Because the ```hashAgg``` operator has two variables and the information displayed is not the same like other variables such as rows. So I think it is better to use string than int, what's your opinion? @lzmhhh123

File: util/execdetails/execdetails.go (hunk 9 above)
Comment (Reminiscent) [reply util/execdetails/execdetails.go]: It should be modified. Otherwise, it can not be compiled.

File: util/execdetails/execdetails.go (hunk 8 above)
Comment (lzmhhh123) [reply util/execdetails/execdetails.go]: Can we format like this?

```golang
concurrency []struct{name string, number int}
```

**Comment by Reminiscent** [comment]
@SunRunAway  @lzmhhh123  Update. PTAL

File: util/execdetails/execdetails.go (hunk 8 above)
Comment (lzmhhh123) [reply util/execdetails/execdetails.go]: Or
```golang
concurrency map[string]int
```

File: executor/aggregate.go [hunk 10]
```diff
@@ -226,11 +226,23 @@ func (e *HashAggExec) Close() error {
...
 
-	// len(e.partialWorkers) == len(e.finalWorkers)
-	concurrency := len(e.partialWorkers)
-	if e.runtimeStats != nil && concurrency > 0 {
+	if e.runtimeStats != nil {
+		var partialInfo, finalInfo string
 		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
-		rootStats.SetConcurrencyInfo(fmt.Sprintf("PartialConcurrency:%d, FinalConcurrency:%d", concurrency, concurrency))
+		partialConcurrency := len(e.partialWorkers)
+		if partialConcurrency > 1 {
```
Comment (SunRunAway) [inline executor/aggregate.go]: It's better to use `e.isUnparallelExec` to determine this judgment.

File: executor/aggregate.go [hunk 11]
```diff
@@ -226,11 +226,23 @@ func (e *HashAggExec) Close() error {
...
+			partialInfo = fmt.Sprintf("PartialConcurrency:%d", partialConcurrency)
+		} else {
+			partialInfo = fmt.Sprintf("PartialConcurrency:OFF")
+		}
+
+		finalConcurrency := len(e.finalWorkers)
+		if finalConcurrency > 1 {
```
Comment (SunRunAway) [inline executor/aggregate.go]: ditto

File: executor/index_lookup_hash_join.go [hunk 12]
```diff
@@ -293,16 +293,20 @@ func (e *IndexNestedLoopHashJoin) Close() error {
 		}
 		e.taskCh = nil
 	}
+	if e.runtimeStats != nil {
+		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
+		concurrency := len(e.joinChkResourceCh)
+		if concurrency > 1 {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", concurrency))
+		} else {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:OFF"))
```
Comment (SunRunAway) [inline executor/index_lookup_hash_join.go]: Even if concurrency is 1, index lookup hash join still invoke a groutine. I think we should not show `OFF` here.

File: executor/index_lookup_join.go [hunk 13]
```diff
@@ -679,5 +679,14 @@ func (e *IndexLookUpJoin) Close() error {
 	}
 	e.workerWg.Wait()
 	e.memTracker = nil
+	if e.runtimeStats != nil {
+		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
+		concurrency := len(e.resultCh)
+		if concurrency > 1 {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", concurrency))
+		} else {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:OFF"))
```
Comment (SunRunAway) [inline executor/index_lookup_join.go]: ditto.

File: executor/index_lookup_merge_join.go [hunk 14]
```diff
@@ -684,5 +684,14 @@ func (e *IndexLookUpMergeJoin) Close() error {
 	}
 	e.joinChkResourceCh = nil
 	e.memTracker = nil
+	if e.runtimeStats != nil {
+		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
+		concurrency := len(e.resultCh)
+		if concurrency > 1 {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", concurrency))
+		} else {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:OFF"))
```
Comment (SunRunAway) [inline executor/index_lookup_merge_join.go]: ditto.

File: executor/join.go [hunk 15]
```diff
@@ -127,10 +127,14 @@ func (e *HashJoinExec) Close() error {
...
-	concurrency := len(e.joiners)
-	if e.runtimeStats != nil && concurrency > 0 {
+	if e.runtimeStats != nil {
 		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
-		rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", concurrency))
+		concurrency := len(e.joiners)
+		if concurrency > 1 {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", concurrency))
+		} else {
+			rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:OFF"))
```
Comment (SunRunAway) [inline executor/join.go]: ditto.

File: executor/projection.go [hunk 16]
```diff
@@ -259,9 +259,13 @@ func (e *ProjectionExec) Close() error {
 		}
 		e.outputCh = nil
 	}
-	if e.runtimeStats != nil && e.numWorkers > 0 {
+	if e.runtimeStats != nil {
 		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
-		rootStats.SetConcurrencyInfo(fmt.Sprintf("Concurrency:%d", e.numWorkers))
+		if e.numWorkers > 1 {
```
Comment (SunRunAway) [inline executor/projection.go]: We may use `e.isUnparallelExec` here.

**Comment by SunRunAway** [comment]
And please resolve this comment https://github.com/pingcap/tidb/pull/13539/files#r349036708

File: util/execdetails/execdetails.go (hunk 2 above)
Comment (Reminiscent) [reply util/execdetails/execdetails.go]: Update. Use the mutex to prevent data race

**Comment by Reminiscent** [comment]
@lzmhhh123  @SunRunAway  Update. PTAL

File: executor/index_lookup_hash_join.go [hunk 17]
```diff
@@ -292,6 +292,11 @@ func (e *IndexNestedLoopHashJoin) Close() error {
 		}
 		e.taskCh = nil
 	}
+	if e.runtimeStats != nil {
+		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
+		concurrency := len(e.joinChkResourceCh)
```
Comment (SunRunAway) [inline executor/index_lookup_hash_join.go]: `len` for a channel means  the number of elements queued (unread) in the channel buffer.
We should use `cap` here.

The same as below.

File: executor/projection.go [hunk 18]
```diff
@@ -259,6 +259,14 @@ func (e *ProjectionExec) Close() error {
 		}
 		e.outputCh = nil
 	}
+	if e.runtimeStats != nil {
+		rootStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.baseExecutor.id.String())
+		if e.isUnparallelExec() {
+			rootStats.SetConcurrencyInfo("Concurrency", int(e.numWorkers))
+		} else {
+			rootStats.SetConcurrencyInfo("Concurrency", 0)
+		}
```
Comment (SunRunAway) [inline executor/projection.go]: Emmm, we should exchange the content of if & else.

File: util/execdetails/execdetails.go [hunk 19]
```diff
@@ -410,6 +421,23 @@ func (e *RuntimeStats) SetRowNum(rowNum int64) {
...
-	return fmt.Sprintf("time:%v, loops:%d, rows:%d", time.Duration(e.consume), e.loop, e.rows)
+	result := fmt.Sprintf("time:%v, loops:%d, rows:%d", time.Duration(e.consume), e.loop, e.rows)
+	if len(e.concurrency) > 0 {
+		for _, concurrency := range e.concurrency {
+			if concurrency.concurrencyNum > 0 {
+				result += fmt.Sprintf(", %s:%d", concurrency.concurrencyName, concurrency.concurrencyNum)
+			} else {
+				result += fmt.Sprintf(", %s:OFF", concurrency.concurrencyName)
+			}
+		}
```
Comment (SunRunAway) [inline util/execdetails/execdetails.go]: We should lock this section.

File: util/execdetails/execdetails.go [hunk 20]
```diff
@@ -318,14 +320,23 @@ type RuntimeStatsColl struct {
...
 
+// concurrencyInfo is used to save the concurrency information of the executor operator
+type concurrencyInfo struct {
+	concurrencyName string
+	concurrencyNum  int
+}
+
 // RuntimeStats collects one executor's execution info.
 type RuntimeStats struct {
+	mu sync.Mutex
```
Comment (SunRunAway) [inline util/execdetails/execdetails.go]: Place `mu` just above the `concurency`, to show that `concurrency` is protected by `mu`.

File: util/execdetails/execdetails.go (hunk 19 above)
Comment (lzmhhh123) [reply util/execdetails/execdetails.go]: Why we should lock this? Does anywhere cause the data race?

Comment (SunRunAway) [reply util/execdetails/execdetails.go]: My fault, it's acceptable not locking this section. Should we add a comment for this method and state that it is not thread safe?

File: util/execdetails/execdetails.go [hunk 21]
```diff
@@ -410,6 +422,25 @@ func (e *RuntimeStats) SetRowNum(rowNum int64) {
 	atomic.StoreInt64(&e.rows, rowNum)
 }
 
+// SetConcurrencyInfo sets the concurrency information.
+func (e *RuntimeStats) SetConcurrencyInfo(name string, num int) {
```
Comment (SunRunAway) [inline util/execdetails/execdetails.go]: We may add an annotation for the situation of `num <= 0`
